wagtail/contrib/search_promotions/views.py

```diff
@@ -1,85 +1,78 @@
-
-    queries = queries.order_by(ordering)
-
-    # Search
-    is_searching = False
-    query_string = request.GET.get("q", "")
-
-    if query_string:
-        queries = queries.filter(query_string__icontains=query_string)
-        is_searching = True
-
-    # Paginate
-    paginator = Paginator(queries, per_page=20)
-    try:
-        queries = paginator.page(request.GET.get("p", 1))
-    except InvalidPage:
-        raise Http404
-
-    if request.headers.get("x-requested-with") == "XMLHttpRequest":
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/results.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-            },
-        )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/index.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                ),
-            },
+class IndexView(generic.IndexView):
+    model = Query
+    template_name = "wagtailsearchpromotions/index.html"
+    results_template_name = "wagtailsearchpromotions/results.html"
+    context_object_name = "queries"
+    page_title = gettext_lazy("Search Terms")
+    any_permission_required = ["add", "change", "delete"]
+    paginate_by = 20
+    page_kwarg = "p"
+    index_url_name = "wagtailsearchpromotions:index"
+    index_results_url_name = "wagtailsearchpromotions:index_results"
+    is_searchable = True
+    default_ordering = "query_string"
+    search_fields = ["promoted_results", "query_string"]
+    _show_breadcrumbs = True
+    add_url_name = "wagtailsearchpromotions:add"
+    breadcrumbs_items = [add_url_name]
+    columns = [
+        TitleColumn(
+            "query_string",
+            classname="query_string",
+            label=gettext_lazy("Search term(s)"),
+            width="40%",
+            url_name="wagtailsearchpromotions:edit",
+            sort_key="query_string",
+        ),
+        Column(
+            "promoted_results",
+            label=gettext_lazy("Promoted results"),
+            width="40%",
+        ),
+        Column(
+            "views",
+            label=gettext_lazy("Views (Past week)"),
+            width="20%",
+            sort_key="views",
+        ),
+    ]
+
+    def get_base_queryset(self):
+        queryset = self.model.objects.filter(editors_picks__isnull=False).distinct()
+        queryset = queryset.annotate(
+            views=functions.Coalesce(Sum("daily_hits__hits"), 0),
+            promoted_results=F("editors_picks__page__title"),
```

Comment:
@laymonage Sorry i am little confused!

```
def get_base_queryset(self):
        queryset = self.model.objects.filter(editors_picks__isnull=False).distinct()
        queryset = queryset.annotate(views=functions.Coalesce(Sum("daily_hits__hits"), 0))
        queryset = queryset.prefetch_related("editors_picks", "editors_picks__page")
        return queryset
```
When using `prefetch_related`, what should I use as the name in `SearchPromotionColumn`? When trying `editors_picks__page__title`, I received an error message:

_`editors_picks__page__title' does not resolve to an item that supports prefetching - this is an invalid parameter to prefetch_related().`_

So, how can i access page titles?

Reply:
The two aren't directly related. You specify `editors_picks` and `editors_picks__page` in `prefetch_related` so that the instances of those models are queried in bulk. Then, you can pass `editors_picks` to the column. This means the `value` variable in the `SearchPromotionColumn` would be the `editors_picks`, i.e. a set of `SearchPromotion` instances. From there you can loop the `value` to get each `SearchPromotion` instance, so essentially in the templates you "just" replace `{% for searchpick in query.editors_picks.all %}` with `{% for searchpick in value %}`.

Reply:
@laymonage Thanks a lot!

I properly understood what you said. But, I am just getting a little issue.





> Then, you can pass `editors_picks` to the column.

As you told to pass the `editors_picks` to the column. But, I think we cannot access the related values (like - Foreignkeys) directly, isn't it?

### Column:
```
SearchPromotionColumn(
            "editors_picks",
            label=_("Promote"),
        ), 

```

Therefore, I am getting an error that says:

`create_reverse_many_to_one_manager.<locals>.RelatedManager.__call__() missing 1 required keyword-only argument: 'manager'`

So, how can i approach this?

Reply:
Does `editors_picks.all` work?

Reply:
@laymonage I tried but still getting the same error.

Reply:
Thankyou, finally it worked 🙌!